src/ffi.rs

```diff
@@ -248,6 +248,52 @@ extern "C" {
     ) -> c_int;
 }
 
+
+#[no_mangle]
+/// **This function is an override for the C function, this is the an edited version of the original description:**
+///
+/// A callback function to be called when an illegal argument is passed to
+/// an API call. It will only trigger for violations that are mentioned
+/// explicitly in the header. **This will cause a panic**.
+///
+/// The philosophy is that these shouldn't be dealt with through a
+/// specific return value, as calling code should not have branches to deal with
+/// the case that this code itself is broken.
+///
+/// On the other hand, during debug stage, one would want to be informed about
+/// such mistakes, and the default (crashing) may be inadvisable.
+/// When this callback is triggered, the API function called is guaranteed not
+/// to cause a crash, though its return value and output arguments are
+/// undefined.
+///
+/// See also secp256k1_default_error_callback_fn.
+///
+pub extern "C" fn secp256k1_default_illegal_callback_fn(_message: *const c_char, _data: *mut c_void) {
+    // Do we need to deref the message and print it? if so without std we'll need to use `strlen`
+    panic!("[libsecp256k1] illegal argument.");
+}
+
+#[no_mangle]
+/// **This function is an override for the C function, this is the an edited version of the original description:**
+///
+/// A callback function to be called when an internal consistency check
+/// fails. **This will cause a panic**.
+///
+/// This can only trigger in case of a hardware failure, miscompilation,
+/// memory corruption, serious bug in the library, or other error would can
+/// otherwise result in undefined behaviour. It will not trigger due to mere
+/// incorrect usage of the API (see secp256k1_default_illegal_callback_fn
+/// for that). After this callback returns, anything may happen, including
+/// crashing.
+///
+/// See also secp256k1_default_illegal_callback_fn.
+///
+pub extern "C" fn secp256k1_default_error_callback_fn(_message: *const c_char, _data: *mut c_void) {
+    // Do we need to deref the message and print it? if so without std we'll need to use `strlen`
```

Comment:
Couldn't we implement our own strlen?
We could also try submit an upstream PR that gives the length to the callback or fixes it to some constant.

Reply:
We could. my question is how important is the message anyway? 
I found some places it's used with "Invalid Scratch space" or checkpoint, but I don't think those should even ever happen and if so then it might be fine to panic without printing?

Reply:
Those *should* never happen but I think if they happen we're really interested why they happen, so I'd prefer to have them.

Reply:
we can make it work under the `std` feature only. then we can use `CStr`

Reply:
I'll implement `strlen` later today. should we panic/unwrap or abort in the callbacks? I see no reason not to unwrap(I see no obvious reason why not to unwrap the stack).

Reply:
Cool. Yep, panic is the right thing to do. (But I don't understand what you mean by unwrap. What can be unwrapped here?)

Reply:
a. panic == unwrap.
b. i'll need to then use `str::from_utf8` and unwrap the result.

Reply:
Okay, sorry, I'm still confused here.  How will this be `no_std` if we call need to call `str::from_utf8`?

re `strlen`: See https://github.com/rust-lang/rust/blob/master/src/libstd/ffi/c_str.rs#L965, `std` seems to just call it and this seems to work in LLVM (https://github.com/rust-lang/rust/issues/29130) but actually I don't know they they do this. But you could try to do the same. :D

Reply:
The whole `str` module is in libcore not libstd, so we can still use this.

And yeah. they have an implementation here: https://github.com/rust-lang/rust/blob/master/src/libstd/sys/sgx/mod.rs#L118

Reply:
Oh sure, nevermind.